File: 01.剑指offer/30.包含min函数的栈/solution.py
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack_data = []
        self.stack_min = []

    def push(self, x: int) -> None:
        self.stack_data.insert(0, x)

        if len(self.stack_min) == 0:
            self.stack_min.insert(0, x)
        else:
            if x <= self.stack_min[0]:
                self.stack_min.insert(0, x)

    def pop(self) -> None:
        val = self.stack_data[0]
        self.stack_data.pop(0)

        if val <= self.stack_min[0]:
            self.stack_min.pop(0)


    def top(self) -> int:
        return self.stack_data[0]

    def min(self) -> int:
        return self.stack_min[0]


if __name__ == "__main__":
    stack = MinStack()
    stack.push(5)
    print(stack.min())
    stack.push(4)
    print(stack.min())
    stack.push(3)
    print(stack.min())
    stack.push(2)
    print(stack.min())

File: 01.剑指offer/30.包含min函数的栈/solution.cpp
//
// Created by lifan on 2022/5/22.
//

#include <stack>
#include <iostream>

class MinStack {
 public:
  /** initialize your data structure here. */
  std::stack<int> _data, _min;

  MinStack() {
  }

  void push(int x) {
      _data.push(x);
      if (_min.empty()) {
          _min.push(x);
      } else if (_min.top() >= x) {
          _min.push(x);
      }
  }

  void pop() {
      if (_min.top() >= _data.top()) {
          _min.pop();
      }
      _data.pop();
  }

  int top() {
      return _data.top();
  }

  int min() {
      return _min.top();
  }
};

int main(int argc, char *argv[])
{
    MinStack s;
    s.push(4);
    std::cout << s.min() << std::endl;
    s.push(3);
    std::cout << s.min() << std::endl;
    s.push(2);
    std::cout << s.min() << std::endl;
    s.push(1);
    std::cout << s.min() << std::endl;

    return 0;
}